merkle_zeppelin/trees/merkle.py

```diff
@@ -16,67 +20,141 @@ def keccak256(v: bytes) -> bytes:
+            cls.validate(obj)
+
+        return obj
+
+    @property
+    def dto(self) -> MerkleTreeDTO:
+        return MerkleTreeDTO(
+            tree=[f"0x{node.hex()}" for node in self._nodes],
+            values=[
+                LeafValueDTO(value=value, treeIndex=index)
+                for value, index in self._raw_to_leaves_index_mapping.items()
+            ],
+            leafEncoding=self._types,
+        )
+
+    def export_tree(self, exporter: Type[MerkleTreeExporter]) -> Any:
+        return exporter.export_tree(self.dto)
 
     def validate(self, raise_exception: bool = True) -> bool:
-        for i in range(1, self._inner_nodes_number):
-            left_node_index = self._get_left_child_index(i)
-            right_node_index = self._get_right_child_index(i)
+        calculated_leaves = sorted(
+            [
+                self._calculate_leaf_hash(value)
+                for value in self._raw_to_leaves_index_mapping.keys()
+            ],
+            reverse=True,
+        )
+        if self.leaves != calculated_leaves:
+            raise MerkleTreeValidationFailed()
+
+        for checked, i in enumerate(range(len(self._nodes) - 1, 0, -2)):
             calculated_parent = self._calculate_parent_value(
-                self._nodes[left_node_index], self._nodes[right_node_index]
+                self._nodes[i], self._nodes[i - 1]
             )
 
-            if self._nodes[i] != calculated_parent:
+            if self._nodes[self._inner_nodes_number - checked - 1] != calculated_parent:
                 if not raise_exception:
                     return False
 
                 raise MerkleTreeValidationFailed()
 
         return True
 
+    # 1470,20 + 4965 = 6445.20
+
     def get_proofs(self, value: Leaf) -> Union[list[bytes], None]:
         try:
-            node_index = self._raw_leaves_index[value]
+            node_index = self._raw_to_leaves_index_mapping[value]
         except ValueError:
             raise ValueNotFoundInTree(value)
 
         result = []
         while node_index != 0:
+            print(node_index)
             neighbor_index = self._get_neighbor_index(node_index)
             result.append(self._nodes[neighbor_index])
 
             node_index = self._get_parent_index(node_index)
 
         return result
 
+    def _get_sorted_index_hash_pairs(
+        self, raw_elements: list[Leaf]
+    ) -> list[tuple[int, bytes]]:
+        hashes = [self._calculate_leaf_hash(el) for el in raw_elements]
+
+        return sorted(
+            enumerate(hashes),
+            reverse=True,
+            key=itemgetter(1),  # sort by hash (second position)
+        )
+
+    def _get_raw_to_leaves_index_mapping(
+        self, raw_elements: list[Leaf], ordered_hashed_leaves: list[tuple[int, bytes]]
+    ) -> dict[Leaf, int]:
+        elements_number = len(raw_elements)
+
+        related_indexed_when_sorted = sorted(
+            range(elements_number), key=ordered_hashed_leaves.__getitem__
```

Comment:
Can we use `itemgetter` here?
```suggestion
            range(elements_number), key=itemgetter(ordered_hashed_leaves)
```
It creates a callable that accepts an object (in this case, an index) and returns the item at that position in the sequence. In your specific context, since `ordered_hashed_leaves` is a list of tuples and you want to sort based on these tuples, `itemgetter` can be used to fetch each tuple based on the indices provided by `range(elements_number)`.

In general, I'd try to avoid using magic methods when possible.

Reply:
unfortunately it does not work as you described :/
anyway, I can agree in this particular case to avoid magic method, it raises more confusion than brings simplicity
I will simply change it to lambda

Reply:
Could you share some details? "Not working" doesn't say to much.